models/staging/ga4/stg_ga4__events.sql

```diff
@@ -13,29 +13,22 @@ add_user_key as (
         *,
         case
             when user_id is not null then to_base64(md5(user_id))
-            when user_pseudo_id is not null then to_base64(md5(user_pseudo_id))
-            else null -- this case is reached when privacy settings are enabled
+            when user_pseudo_id is not null and user_pseudo_id != '' then to_base64(md5(user_pseudo_id))
+            else null -- this case is reached when privacy settings are enabled and (possibly) for certain audience triggered events
         end as user_key
     from base_events
 ), 
 -- Add unique keys for sessions and events
 include_session_key as (
     select 
         *,
-        to_base64(md5(CONCAT(stream_id, user_pseudo_id, ga_session_id))) as session_key -- Surrogate key to determine unique session across streams and users. Sessions do NOT reset after midnight in GA4
+        to_base64(md5(CONCAT(stream_id, CAST(user_key as STRING), IFNULL(CAST(ga_session_id as STRING), 'no_session')))) as session_key -- Surrogate key to determine unique session across streams and users. Sessions do NOT reset after midnight in GA4
```

Comment:
Is the revert back to using user_key instead of user_pseudo_id deliberate here?

I don't think the CASTs are necessary BTW, aren't all those things strings? I don't remember ever seeing a NULL ga_session_id either personally.

Reply:
Right, use of user_key was an oversight. I'll update that. 

`ga_session_id` can definitely be `null`. See audience triggers: https://support.google.com/analytics/answer/9934109?hl=en